### PR TITLE
Fixed Issue/#2216

### PR DESCRIPTION
**Pull request by NayanKhedkar**
Added new  functionality to track state on language switch.

### CODE REVIEW

File: js/languagePickerModel.js [hunk 1]
```diff
@@ -45,6 +55,81 @@ define([
...
+            var blocks = this.trackedData.blocks;
+
+            if (components) {
+                components.forEach(this.setTrackableState);
+            }
+
+            if (blocks) {
+                blocks.forEach(this.setTrackableState);
+            }
+        },
```
Comment (moloko) [inline js/languagePickerModel.js]: local variables seem a little unnecessary here; could just be written as 
```js
if (this.trackedData.components) {
    this.trackedData.components.forEach(this.setTrackableState);
}

if (this.trackedData.blocks) {
    this.trackedData.blocks.forEach(this.setTrackableState);
}
```

File: js/languagePickerModel.js [hunk 2]
```diff
@@ -45,6 +55,81 @@ define([
...
+                this.restoreState();
+            }.bind(this));
+
+        },
+
+        restoreLocation: function() {
+            if (!Adapt.mapById(this.locationId)) return;
+            _.defer(function() {
+                Adapt.navigateToElement('.' + this.locationId);
+            }.bind(this));
```
Comment (moloko) [inline js/languagePickerModel.js]: slightly tidier if written as:
```js
_.defer(Adapt.navigateToElement, '.' + this.locationId);
```

File: properties.schema [hunk 3]
```diff
@@ -88,6 +88,15 @@
                   "validators": [],
                   "help": "Your default options here are: icon-globe, icon-language-1, icon-language-2"
                 },
+                "_restoreStateOnLanguageChange": {
+                  "type": "boolean",
+                  "required": true,
+                  "default": true,
```
Comment (brian-learningpool) [inline properties.schema]: I think the default should be `false` to maintain backwards compatibility with existing courses.

File: properties.schema [hunk 4]
```diff
@@ -88,6 +88,15 @@
...
+                  "title": "Restore course state.",
```
Comment (brian-learningpool) [inline properties.schema]: No need for a full-stop here.

File: properties.schema [hunk 5]
```diff
@@ -88,6 +88,15 @@
...
+                  "inputType": "Checkbox",
+                  "validators": [],
+                  "help": "Restore course state on language switch."
```
Comment (brian-learningpool) [inline properties.schema]: I think the `help` text should give some mention that it's strongly recommended that the course structure matches among languages, e.g. for future proofing the authoring tool, something like:
`NOTE: It's recommended that courses share the same structure.`

**Comment by brian-learningpool** [comment]
Also, can you update the framework dependency to v3+?

File: js/languagePickerModel.js [hunk 6]
```diff
@@ -11,8 +11,18 @@ define([
...
+        trackedData: {
+            "components": [],
+            "blocks": []
+        },
+
+        locationId: null,
+
         initialize: function () {
             this.listenTo(Adapt.config, 'change:_activeLanguage', this.onConfigChange);
+            if (this.get('_restoreStateOnLanguageChange')) {
```
Comment (moloko) [inline js/languagePickerModel.js]: you can get rid of this check now i.e. always listen to the `app:dataLoaded` event

**Comment by moloko** [comment]
> Also, can you update the framework dependency to v3+?

FW dependency is already set to ">=3"

**Pull request merged**